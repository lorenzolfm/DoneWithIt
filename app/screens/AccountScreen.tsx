import React from 'react';
import { ListItem } from '../components/ListItem';

import { Screen } from '../components/Screen';

export const AccountScreen = () => {
  return (
    <Screen>
      <ListItem
        title="Mosh Hamedani"
        subtitle="programmingwithmosh@gmail.com"
      />
    </Screen>
  );
};
