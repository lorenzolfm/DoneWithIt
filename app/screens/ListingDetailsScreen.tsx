import React from 'react';
import { Image, View, StyleSheet } from "react-native";

import { ListItem } from '../components/ListItem';
import { AppText } from "../components/AppText";

import defaultStyles from '../config/styles'

export const ListingDetailsScreen = ({ route }) => {
  const listing = route.params;

  return (
    <View>
      <Image style={styles.image} source={listing.image} />
      <View style={styles.detailsContainer}>
        <AppText style={styles.title}>{listing.title}</AppText>
        <AppText style={styles.price}>{listing.price}</AppText>
        <View style={styles.userContainer}>
          <ListItem
            image={require('../assets/mosh.jpg')}
            title="Mosh Hamedani"
            subtitle="5 Listings"
          />
        </View>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  image: {
    width: '100%',
    height: 300,
  },
  detailsContainer: {
    padding: 20,
  },
  title: {
    fontSize: 24,
    fontWeight: '500',
  },
  price: {
    color: defaultStyles.colors.secondary,
    fontWeight: 'bold',
    fontSize: 20,
    marginVertical: 10,
  },
  userContainer: {
    marginVertical: 40,
  }
});
