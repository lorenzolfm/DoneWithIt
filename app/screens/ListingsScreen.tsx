import React, { useEffect } from 'react';
import { FlatList, StyleSheet } from "react-native";

import { Card } from '../components/Card';
import { Screen } from "../components/Screen";
import { ActivityIndicator } from '../components/ActivityIndicator';
import { Navigation } from '../types'
import routes from '../navigation/routes';
import { getListings } from '../api/listings';

import defaultStyles from '../config/styles';
import { AppText } from '../components/AppText';
import { Button } from '../components/Button';
import { useApi } from '../hooks/useApi';

export const ListingsScreen = ({ navigation: { navigate } }: Navigation) => {
  const { data: listings, error, loading, request: loadListings } = useApi(getListings)

  useEffect(() => {
    loadListings();
  }, [])

  return (
    <Screen style={styles.screen}>
      {error && <>
        <AppText>Couldn't retrieve listings</AppText>
        <Button onPress={loadListings} title="Retry" />
      </>}
      <ActivityIndicator visible={loading}/>
      <FlatList
        data={listings}
        keyExtractor={listing => listing.id.toString()}
        renderItem={({ item }) =>
          <Card
            title={item.title}
            subtitle={"$" + item.price}
            imageUrl={item.images[0].url}
            thumbnailUrl={item.images[0].thumbnailUrl}
            onPress={() => navigate(routes.LISTING_DETAILS, item)}
          />
        }
      />
    </Screen>
  );
};

const styles = StyleSheet.create({
  screen: {
    padding: 20,
    backgroundColor: defaultStyles.colors.light,
  },
});
