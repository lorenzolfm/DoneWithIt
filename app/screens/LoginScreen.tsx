import React from 'react';
import { Image, StyleSheet } from 'react-native';
import { Formik } from 'formik';
import * as Yup from 'yup';

import { AppTextInput } from '../components/AppTextInput';
import { Button } from '../components/Button';
import { Screen } from '../components/Screen';
import { AppText } from '../components/AppText';

const validationSchema = Yup.object().shape({
  email: Yup.string().required().email().label('Email'),
  password: Yup.string().required().min(4).label('Password'),
});

const initialValues = {
  email: '',
  password: '',
};

export const LoginScreen = () => {
  return (
    <Screen style={styles.container}>
      <Image source={require("../assets/logo-red.png")} style={styles.logo}/>
      <Formik
        initialValues={initialValues}
        onSubmit={values => console.log(values)}
        validationSchema={validationSchema}
      >
        { ({ handleChange, handleSubmit, errors }) => (
          <>
            <AppTextInput
              autoCapitalize="none"
              autoCorrect={false}
              keyboardType="email-address"
              onChangeText={handleChange('email')}
              icon="email"
              placeholder="Email"
              textContentType="email"
            />
            <AppText style={{ color: 'red' }}>{errors.email}</AppText>
            <AppTextInput
              autoCapitalize="none"
              autoCorrect={false}
              icon="lock"
              onChangeText={handleChange('password')}
              placeholder="Password"
              textContentType="password"
              secureTextEntry
            />
            <AppText style={{ color: 'red' }}>{errors.password}</AppText>
            <Button title="Login" onPress={() => handleSubmit()} />
          </>
        )}
      </Formik>
    </Screen>
  );
};

const styles = StyleSheet.create({
  container: {
    padding: 10,
  },
  logo: {
    width: 80,
    height: 80,
    alignSelf: 'center',
    marginTop: 50,
    marginBottom: 20,
  },
});
