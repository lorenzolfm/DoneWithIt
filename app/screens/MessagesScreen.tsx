import React, { useState } from 'react';
import { FlatList } from "react-native"

import { ListItem } from '../components/ListItem';
import { ListItemDeleteAction } from '../components/ListItemDeleteAction';
import { ListItemSeparator } from '../components/ListItemSeparator';
import { Screen } from '../components/Screen';

interface Message {
  id: number,
  title: string,
  description: string,
  image: any,
};

const msgs: Array<Message> = [
  {
    id: 0,
    title: 'T1',
    description: 'D1',
    image: require('../assets/mosh.jpg'),
  },
  {
    id: 1,
    title: 'T2',
    description: 'D2',
    image: require('../assets/mosh.jpg'),
  },
];

export const MessagesScreen = () => {
  const [messages, setMessages] = useState(msgs);
  const [refreshing, setRefreshing] = useState(false);

  const handleDelete = (message: Message): void => {
    setMessages(messages.filter(m => m.id !== message.id));
  };

  return (
    <Screen>
      <FlatList
        data={messages}
        keyExtractor={item => item.id.toString()}
        renderItem={
          ({ item }) =>
            <ListItem
              title={item.title}
              subtitle={item.description}
              image={item.image}
              onPress={() => console.log('Message selected', item)}
              renderRightActions={() => <ListItemDeleteAction onPress={() => handleDelete(item)} />}
            />
        }
        ItemSeparatorComponent={ListItemSeparator}
        refreshing={refreshing}
        onRefresh={() => setMessages([...msgs])}
      />
    </Screen>
  );
};
