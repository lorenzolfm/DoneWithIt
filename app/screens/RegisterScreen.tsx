import React, { useState } from 'react';
import { StyleSheet } from 'react-native';
import * as Yup from 'yup';

import { Screen } from '../components/Screen';
import { ErrorMessage, AppForm, AppFormField, SubmitButton } from '../components/Forms';
import { User } from '../types';
import { register } from '../api/users';
import { login } from '../api/auth';
import { useAuth } from '../auth/useAuth';
import { useApi } from '../hooks/useApi';
import { ActivityIndicator } from '../components/ActivityIndicator';

const validationSchema = Yup.object().shape({
  name: Yup.string().required().label('Name'),
  email: Yup.string().required().email().label('Email'),
  password: Yup.string().required().min(4).label('Password'),
});

const initialValues = { name: '', email: '', password: '' };

export const RegisterScreen = () => {
  const registerApi = useApi(register);
  const loginApi = useApi(login);
  const { logIn } = useAuth();
  const [error, setError] = useState('');

  const handleSubmit = async (userInfo: User): Promise<void> => {
    const result = await registerApi.request(userInfo);

    if (!result.ok) {
      if (result.data) setError(result.data?.error);
      else {
        setError('An unexpected error occurred.');
        console.log(result);
      }
      return;
    }

    const { data: authToken } = await loginApi.request(
      userInfo.email,
      userInfo.password,
    );
    logIn(authToken as string);
  }

  return (
    <>
      <ActivityIndicator visible={registerApi.loading || loginApi.loading} />
      <Screen style={styles.container}>
        <ErrorMessage visible={error ? true : false} error={error as string} />
        <AppForm
          initialValues={initialValues}
          onSubmit={handleSubmit}
          validationSchema={validationSchema}
        >
          <AppFormField
            autoCorrect={false}
            icon="account"
            name="name"
            placeholder="Name"
          />
          <AppFormField
            autoCapitalize="none"
            autoCorrect={false}
            icon="email"
            keyboardType="email-address"
            name="email"
            placeholder="Email"
            textContentType="emailAddress"
          />
          <AppFormField
            autoCapitalize="none"
            autoCorrect={false}
            icon="lock"
            name="password"
            placeholder="Password"
            secureTextEntry
            textContentType="password"
          />
          <SubmitButton title="Register" />
        </AppForm>
      </Screen>
    </>
  );
};

const styles = StyleSheet.create({
  container: {
    padding: 10,
  },
});
