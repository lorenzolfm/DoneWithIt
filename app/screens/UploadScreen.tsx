import React from 'react';
import { View, Modal, StyleSheet } from "react-native"
import { AppText } from "../components/AppText"

type Props = { progress: number, visible: boolean }

export const UploadScreen = ({ progress = 0, visible = false }: Props) => {
  return (
    <Modal visible={visible}>
      <View style={styles.container}>
        <AppText>{progress * 100}%</AppText>
      </View>
    </Modal>
  );
};

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    flex: 1,
    justifyContent: 'center',
  },
});
