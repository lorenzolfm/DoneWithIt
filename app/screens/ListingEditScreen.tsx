import React, { useState } from 'react';
import { StyleSheet } from 'react-native';
import * as Yup from 'yup';
import { useLocation } from '../hooks/useLocation';
import { addListing } from '../api/listings'

import {
  AppForm,
  AppFormField,
  AppFormPicker,
  SubmitButton
} from '../components/Forms'
import { CategoryPickerItem } from '../components/CategoryPickerItem';
import { Screen } from '../components/Screen';
import { FormImagePicker } from '../components/Forms/FormImagePicker';
import { Listing } from '../types';
import { UploadScreen } from './UploadScreen';

const validationSchema = Yup.object().shape({
  title: Yup.string().required().min(1).label('Title'),
  price: Yup.number().required().min(1).max(10000).label('Price'),
  description: Yup.string().label('description'),
  category: Yup.object().required().nullable().label('Category'),
  images: Yup.array().min(1, "Please select at least one image"),
});

const categories = [
  { label: 'Furniture', value: 1, backgroundColor: 'red', icon: 'apps' },
  { label: 'Clothing', value: 2, backgroundColor: 'green', icon: 'email'},
  { label: 'Camera', value: 3, backgroundColor: 'blue', icon: 'lock' },
];

const initialValues = {
  title: '',
  price: '',
  description: '',
  category: null,
  images: [],
};

export const ListingEditScreen = () => {
  const location = useLocation();
  const [uploadVisible, setUploadVisible] = useState<boolean>(false);
  const [progress, setProgress] = useState<number>(0);

  const handleSubmit = async (listing: Listing) => {
    setProgress(0)
    setUploadVisible(true);
    const result = await addListing(
      {...listing, location},
      progress => setProgress(progress)
    );
    setUploadVisible(false);

    if (!result.ok) {
      setUploadVisible(false);
      console.log(result.problem)
      return alert("Could not save the listing");
    }

    alert('Success');
  }

  return (
    <Screen style={styles.container}>
      <UploadScreen progress={progress} visible={uploadVisible} />
      <AppForm
        initialValues={initialValues}
        onSubmit={handleSubmit}
        validationSchema={validationSchema}
      >
        <FormImagePicker name="images" />
        <AppFormField maxLength={255} name="title" placeholder="Title" />
        <AppFormField
          keyboardType="numeric"
          maxLength={8}
          name="price"
          placeholder="Price"
          width={120}
        />
        <AppFormPicker
          items={categories}
          numberOfColumns={3}
          name="category"
          PickerItemComponent={CategoryPickerItem}
          placeholder="Category"
          width="50%"
        />
        <AppFormField
          maxLength={255}
          multiline
          name="description"
          numberOfLines={3}
          placeholder="Description"
        />
        <SubmitButton title="Post"/>
      </AppForm>
    </Screen>
  );
};

const styles = StyleSheet.create({
  container: {
    width: '95%',
    alignSelf: 'center',
    marginTop: 10,
  },
})
