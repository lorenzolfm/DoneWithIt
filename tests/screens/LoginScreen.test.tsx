import { create } from 'react-test-renderer';
import { LoginScreen } from '../../app/screens/LoginScreen';

describe('LoginScreen', () => {
  it('snapshot matches', () => {
    expect(create(<LoginScreen />)).toMatchSnapshot();
  });
});
