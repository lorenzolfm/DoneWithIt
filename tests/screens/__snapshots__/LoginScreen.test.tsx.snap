// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen snapshot matches 1`] = `
<Screen
  style={
    Object {
      "padding": 10,
    }
  }
>
  <Image
    source={1}
    style={
      Object {
        "alignSelf": "center",
        "height": 80,
        "marginBottom": 20,
        "marginTop": 50,
        "width": 80,
      }
    }
  />
  <AppForm
    initialValues={
      Object {
        "email": "",
        "password": "",
      }
    }
    onSubmit={[Function]}
    validationSchema={
      ObjectSchema {
        "_blacklist": ReferenceSet {
          "list": Set {},
          "refs": Map {},
        },
        "_blacklistError": undefined,
        "_excludedEdges": Array [],
        "_nodes": Array [
          "password",
          "email",
        ],
        "_sortErrors": [Function],
        "_typeError": [Function],
        "_whitelist": ReferenceSet {
          "list": Set {},
          "refs": Map {},
        },
        "_whitelistError": undefined,
        "conditions": Array [],
        "deps": Array [],
        "exclusiveTests": Object {},
        "fields": Object {
          "email": StringSchema {
            "_blacklist": ReferenceSet {
              "list": Set {},
              "refs": Map {},
            },
            "_blacklistError": undefined,
            "_typeError": [Function],
            "_whitelist": ReferenceSet {
              "list": Set {},
              "refs": Map {},
            },
            "_whitelistError": undefined,
            "conditions": Array [],
            "deps": Array [],
            "exclusiveTests": Object {
              "email": false,
              "required": true,
            },
            "spec": Object {
              "abortEarly": true,
              "label": "Email",
              "nullable": false,
              "presence": "required",
              "recursive": true,
              "strict": false,
              "strip": false,
            },
            "tests": Array [
              [Function],
              [Function],
            ],
            "transforms": Array [
              [Function],
            ],
            "type": "string",
          },
          "password": StringSchema {
            "_blacklist": ReferenceSet {
              "list": Set {},
              "refs": Map {},
            },
            "_blacklistError": undefined,
            "_typeError": [Function],
            "_whitelist": ReferenceSet {
              "list": Set {},
              "refs": Map {},
            },
            "_whitelistError": undefined,
            "conditions": Array [],
            "deps": Array [],
            "exclusiveTests": Object {
              "min": true,
              "required": true,
            },
            "spec": Object {
              "abortEarly": true,
              "label": "Password",
              "nullable": false,
              "presence": "required",
              "recursive": true,
              "strict": false,
              "strip": false,
            },
            "tests": Array [
              [Function],
              [Function],
            ],
            "transforms": Array [
              [Function],
            ],
            "type": "string",
          },
        },
        "spec": Object {
          "abortEarly": true,
          "nullable": false,
          "presence": "optional",
          "recursive": true,
          "strict": false,
          "strip": false,
        },
        "tests": Array [],
        "transforms": Array [
          [Function],
        ],
        "type": "object",
      }
    }
  >
    <AppFormField
      autoCapitalize="none"
      autoCorrect={false}
      icon="email"
      keyboardType="email-address"
      name="email"
      placeholder="Email"
      textContentType="emailAddress"
    />
    <AppFormField
      autoCapitalize="none"
      autoCorrect={false}
      icon="lock"
      name="password"
      placeholder="Password"
      secureTextEntry={true}
      textContentType="password"
    />
    <SubmitButton
      title="Login"
    />
  </AppForm>
</Screen>
`;
